webapp/src/snippetBuilder.tsx

```diff
@@ -0,0 +1,339 @@
+/// <reference path="../../built/pxtlib.d.ts" />
+
+import * as React from "react";
+import * as data from "./data";
+import * as sui from "./sui";
+import * as md from "./marked";
+import * as compiler from './compiler';
+
+type ISettingsProps = pxt.editor.ISettingsProps;
+
+export interface CreateSnippetBuilderState {
+    visible?: boolean;
+    output?: string;
+    projectView?: pxt.editor.IProjectView;
+    answers?: any; // Will be typed once more clearly defined
+    currentQuestion: number;
+    defaults: any; // Will be typed once more clearly defined
+    mainWorkspace?: Blockly.Workspace;
+    config?: any; // Will be a config type
+}
+
+export class CreateSnippetBuilder extends data.Component<ISettingsProps, CreateSnippetBuilderState> {
+    constructor(props: ISettingsProps) {
+        super(props);
+        this.state = {
+            visible: false,
+            answers: {},
+            currentQuestion: 0, // Index to track current question
+            defaults: {},
+            config: staticConfig,
+            output: staticConfig.output
+        };
+
+        this.hide = this.hide.bind(this);
+        this.cancel = this.cancel.bind(this);
+        this.confirm = this.confirm.bind(this);
+        this.backPage = this.backPage.bind(this);
+        this.nextPage = this.nextPage.bind(this);
+    }
+
+    buildDefaults() {
+        const { config } = this.state;
+        const defaults: any = {};
+
+        for (const question of config.questions) {
+            const { inputs } = question;
+            for (const input of inputs) {
+                const { defaultAnswer, answerToken } = input;
+                defaults[answerToken] = defaultAnswer;
+            }
+        }
+
+        this.setState({ defaults });
+    }
+
+    componentDidMount() {
+        // Sets default values
+        this.buildDefaults();
+    }
+
+    replaceTokens(output: string) {
+        const { answers, defaults } = this.state;
+        let tokenizedOutput = output;
+        const tokens = Object.keys(defaults);
+
+        // Replaces output tokens with answer if available or default value
+        for (let token of tokens) {
+            if (answers[token]) {
+                tokenizedOutput = tokenizedOutput.split(`$${token}`).join(answers[token]);
+            }
+            else {
+                tokenizedOutput = tokenizedOutput.split(`$${token}`).join(defaults[token]);
```

Comment:
I think this if statement could become:
```
const value = answers[token] || defaults[token]
tokenizedOutput = tokenizedOutput.split(`$${token}`).join(value);

```

Reply:
do it in 1 pass: ``tokenizedOutput.replace(/\$(\w+)/g, (m, token) => answers[token] || defaults[token])``